src/main/haskell/kore/src/Kore/Step/Simplification/AndTerms.hs

```diff
@@ -627,52 +633,65 @@ equalInjectiveHeadsAndEquals
     -- ^ Used to simplify subterm "and".
     -> PureMLPattern level variable
     -> PureMLPattern level variable
-    -> Result (m (ExpandedPattern level variable, SimplificationProof level))
+    -> MaybeT m (ExpandedPattern level variable, SimplificationProof level)
 equalInjectiveHeadsAndEquals
     tools
     substitutionSimplifier
     termMerger
     (App_ firstHead firstChildren)
     (App_ secondHead secondChildren)
-  | StepperAttributes.isInjective firstHeadAttributes
-    && StepperAttributes.isInjective secondHeadAttributes
-    && firstHead == secondHead
-  = do -- Result monad
-    intCounterChildren <- sequenceA $
-        zipWith termMerger firstChildren secondChildren
-    return $ do -- Counter monad
-        children <- sequenceA intCounterChildren
-        (   PredicateSubstitution
-                { predicate = mergedPredicate
-                , substitution = mergedSubstitution
-                }
-            , _proof
-            ) <- mergePredicatesAndSubstitutions
-                tools
-                substitutionSimplifier
-                (map (ExpandedPattern.predicate . fst) children)
-                (map (ExpandedPattern.substitution . fst) children)
-        return
-            ( Predicated
-                { term = give (MetadataTools.symbolOrAliasSorts tools) $
-                    mkApp firstHead (map (ExpandedPattern.term . fst) children)
-                , predicate = mergedPredicate
-                , substitution = mergedSubstitution
-                }
-            , SimplificationProof
-            )
+  | isFirstInjective && isSecondInjective && firstHead == secondHead
+  = Monad.Trans.lift $ do
+    children <- Monad.zipWithM termMerger firstChildren secondChildren
+    let predicates = ExpandedPattern.predicate . fst <$> children
+        substitutions = ExpandedPattern.substitution . fst <$> children
+    (merged, _proof) <-
+        mergePredicatesAndSubstitutions
+            tools
+            substitutionSimplifier
+            predicates
+            substitutions
+    let PredicateSubstitution
+            { predicate = mergedPredicate
+            , substitution = mergedSubstitution
+            }
+          =
+            merged
+    return
+        ( Predicated
```

Comment:
This is not related to this PR, but could you add a TODO to somehow solve the case when one of the children terms is bottom? I think we don't want to create an application pattern. Most likely this will not break our code since subsequent simplifications will make everything bottom, but it will be confusing, at least. This may also happen in other places.

Reply:
That will also require that it is a symbol, not alias, application, right? I don't know what it would mean to have an injective alias, but it's possible to give that attribute to an alias.